airbyte-workers/src/main/java/io/airbyte/workers/internal/AirbyteMessageTracker.java

```diff
@@ -386,4 +416,54 @@ protected Long calculateMean(final Long currentMean, final Long totalCount, fina
     return (long) result;
   }
 
+  void addStateMessageToStreamToStateHashTimestampTracker(final AirbyteStateMessage stateMessage, final int stateHash, final DateTime timeEmitted) {
+    final List<StreamDescriptor> streamDescriptorsToUpdate = StateMessageHelper.getStreamDescriptors(stateMessage);
+
+    streamDescriptorsToUpdate.forEach(streamDescriptor -> {
+      final HashMap<Integer, DateTime> stateHashToTimestamp = new HashMap<>();
+      stateHashToTimestamp.put(stateHash, timeEmitted);
+      streamDescriptorToStateMessageTimestamps.put(streamDescriptor, stateHashToTimestamp);
+    });
+  }
+
+  Long calculateSecondsBetweenStateEmittedAndCommitted(final DateTime stateMessageEmittedAt, final DateTime stateMessageCommittedAt) {
+    return Long.valueOf(Seconds.secondsBetween(stateMessageEmittedAt, stateMessageCommittedAt).getSeconds());
+  }
+
+  void updateTimestampTrackerAndCalculateMaxAndMeanTimeToCommit(final AirbyteStateMessage stateMessage, final int stateHash) {
+    final DateTime timeCommitted = DateTime.now();
+    final List<StreamDescriptor> streamDescriptorsToResolve = StateMessageHelper.getStreamDescriptors(stateMessage);
+
+    streamDescriptorsToResolve.forEach(streamDescriptor -> {
+      final Map<Integer, DateTime> stateMessagesForStream = streamDescriptorToStateMessageTimestamps.get(streamDescriptor);
+      final DateTime maxTime = stateMessagesForStream.get(stateHash);
+      DateTime minTime = null;
+
+      // update minTime to earliest timestamp that exists for a state message for this particular stream
+      // and delete state message entries that are equal to or earlier than the destination state message
+      for (final Map.Entry<Integer, DateTime> stateMessageTime : stateMessagesForStream.entrySet()) {
```

Comment:
OPT: Looking at this loop, it feels like we may as well track the <hash, datetime> in a queue.
In the current implementation, even though we have direct access through the hash, we have to scans through all the items to expire older messages.
Feels like having a list ordered by time more effective overall.

Reply:
I think we'd still have to iterate through the whole list though and check the timestamp of each to see when we've reached the correct timestamp -- or are you thinking of a different kind of implementation that's not occurring to me right now?

Reply:
If the list is ordered, we can stop as soon as we find a match of the hash. All the items we saw should be expired/removed. Depending on the number of states we end up having in flight, it may matter.

If I got it right, current expectation is that for a given stream, state messages order is preserved, so if we get a state from the destination, it means that everything until that state has been processed. With that in mind, it would be more robust to just keep them in order rather than rely on timestamp for that.

Reply:
ah i see, so still O(n) but in practice it could be faster. so it would look like this:

stream_1: [[state_hash_1, timestamp_1], [state_hash_2, timestamp_2], [state_hash_3, timestamp_3]],
stream_2: [[etc etc]]

then when we get a state message from the destination, we can safely start at the beginning of the state hash list and use the first timestamp as the min timestamp. then we'll remove them one by one as we iterate through until we get to the state hash that matches.